bpf-common/src/program.rs

```diff
@@ -323,21 +332,17 @@ impl Program {
     /// Watch a BPF_MAP_TYPE_PERF_EVENT_ARRAY and forward all its events to `sender`.
     /// A different task is run for each CPU.
     pub async fn read_events<T: Send>(
-        &self,
+        &mut self,
         map_name: &str,
         sender: impl BpfSender<T>,
     ) -> Result<(), ProgramError> {
-        let mut perf_array: AsyncPerfEventArray<_> = self.bpf.map_mut(map_name)?.try_into()?;
-        let maps = online_cpus()
-            .unwrap()
-            .into_iter()
-            .map(|cpu_id| {
-                perf_array
-                    .open(cpu_id, Some(self.ctx.perf_pages))
-                    .map_err(ProgramError::from)
-            })
-            .collect::<Result<Vec<_>, ProgramError>>()?;
-        for mut buf in maps {
+        let map_resource = self
+            .bpf
+            .take_map(map_name)
+            .ok_or_else(|| ProgramError::MapNotFound(map_name.to_string()))?;
+        let mut perf_array: AsyncPerfEventArray<_> = AsyncPerfEventArray::try_from(map_resource)?;
+        for cpu_id in online_cpus().unwrap() {
+            let mut buf = perf_array.open(cpu_id, Some(self.ctx.perf_pages))?;
```

Comment:
apart from the api change , I like the old first-open-all-maps way more. In particular no tasks were started if there was an error opening a map. It's a suggestion , not a blocker